contracts/interfaces/SpokeAdapterInterface.sol

```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+/**
+ * @notice Sends cross chain tokens to from SpokePool to HubPool.
+ */
+
+interface SpokeAdapterInterface {
+    function bridgeTokensToHubPool(uint256 amountToReturn, address l2TokenAddress) external;
```

Comment:
This is really cool! I think this is a huge reduction in the chain-specific lines in each SpokePool implementation.

Do you think it's possible to remove _all_ chain-specific functionality from the SpokePools?

I was thinking that it would be really nice to be able to deploy the exact same SpokePool implementation to each chain along with an adapter that contains all the specialized functionality for: receiving messages from the hub, sending tokens to the HubPool, and anything else that might be specialized based on the chain.

To do this and handle _all_ necessary functionality, the adapter will have to _receive_ a lot of chain-specific function calls. I think this would ultimately either look like one of the two following architectures:
1. All the common SpokePool methods, like deposits and fills will have implementations in the main SpokePool. All specialized or customized functionality would live in the adapter. Any call that hits the SpokePool would be forwarded (via delegatecall) to the adapter by the fallback function if the SpokePool didn't implement it. Note: this is very very tricky from a permissioning perspective, since any methods on the adapter that are _only_ meant to be called by the SpokePool would need to be protected from unauthorized forwards.

2. A full-on proxy where the entire SpokePool implementation sits behind a Proxy, and that implementation can be swapped out by governance. This is a more common pattern and would probably be safer, but gives governance _full_ control over the implementation rather than just controlling some methods.

Thoughts on this general idea? Note: this gets complicated quickly, so I could also see us sticking with only making the token bridging upgradable.

@nicholaspai @pxrl

Reply:
After our discussion offline I think we should favor going with a full [upgradeable proxy pattern](https://docs.openzeppelin.com/upgrades). The [UUPS](https://docs.openzeppelin.com/upgrades-plugins/1.x/#proxy-patterns) seems like a good one for our use case. We'd get the benefit of a well supported and tested proxy with less code change than this PR which refactors the contracts themselves to only have a tiny surface area be upgradeable.

It seems like the main change we'd have to make is to remove the constructor in favor of an `initializer` function, and generally inherit the `Intializable` contract

Reply:
100% agreed! I also think this makes the spoke pool upgradability work _much much_ lower lift and less to audit. +1